Scripts/Data/Collection/ObservableCounter.cs

```diff
@@ -0,0 +1,114 @@
+﻿namespace VRTK.Core.Data.Collection
+{
+    using UnityEngine;
+    using UnityEngine.Events;
+    using System.Linq;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Allows observing counting the amount of attempts an element is added or removed from a <see cref="Dictionary{TKey, TValue}"/>.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the elements to count.</typeparam>
+    /// <typeparam name="TEvent">The <see cref="UnityEvent"/> type to use.</typeparam>
+    public class ObservableCounter<TElement, TEvent> : MonoBehaviour where TEvent : UnityEvent<TElement>, new()
+    {
+        /// <summary>
+        /// Emitted when a searched element is found.
+        /// </summary>
+        public TEvent ElementAdded = new TEvent();
+        /// <summary>
+        /// Emitted when a searched element is not found.
+        /// </summary>
+        public TEvent ElementRemoved = new TEvent();
+
+        /// <summary>
+        /// The elements being counted.
+        /// </summary>
+        public Dictionary<TElement, int> ElementsCounter
+        {
+            get;
+            protected set;
+        } = new Dictionary<TElement, int>();
+
+        /// <summary>
+        /// Increases the count on the given element.
+        /// </summary>
+        /// <param name="element">The element to count.</param>
+        public virtual void IncreaseCount(TElement element)
+        {
+            if (!isActiveAndEnabled || element == null)
+            {
+                return;
+            }
+
+            if (!ElementsCounter.ContainsKey(element))
+            {
+                ElementsCounter.Add(element, 0);
```

Comment:
Remove the `if` and use this instead:
```suggestion
                ElementsCounter[element] = 0;
```
This indexer actually adds a key **or replaces** if the key already exists.

Reply:
That will always set the counter to 0 every time you call IncreaseCount, so count can only ever be 0 or 1